ISPN-6343 Domain mode support for server test suite

mgencur opened the pull request:
https://issues.jboss.org/browse/ISPN-6343

This PR adds a new profile which is activated by passing -Ddomain  
The test suite with this parameter will run all tests which support domain mode. These are currently the "client" tests (basic HotRod, Memcached, REST client tests)

The test suite for domain mode has to run in a separate profile because Arquillian requires either "domain" or "standalone" dependencies on classpath, not both. It fails otherwise. This would be solved by Arquillian-Chameleon project. However, it doesn't support domain mode yet. I filed https://github.com/arquillian/arquillian-container-chameleon/issues/43 

This PR includes workarounds for https://issues.jboss.org/browse/ISPN-6310 and  https://issues.jboss.org/browse/ISPN-6321 which I filed in the meantime.

This PR is the first step that migrates a few basic tests. The rest of the test suite might be ported to the  domain mode later. This will require more work.

slaskawi commented:
For some reason CI missed this PR. Retriggering it: http://ci.infinispan.org/viewQueued.html?itemId=37266

slaskawi commented on server/integration/testsuite/pom.xml:
```diff
@@ -23,26 +23,24 @@
-            <artifactId>wildfly-arquillian-container-managed</artifactId>
+            <groupId>org.infinispan.arquillian.container</groupId>
+            <artifactId>infinispan-arquillian-impl</artifactId>
             <exclusions>
```
Why we are excluding those artifacts?

mgencur replied:
I'm adding it back in a separate profile below. They can't be in "default" profile because they would conflict with domain mode dependencies.

slaskawi replied:
Ok, understood.

slaskawi commented on server/integration/testsuite/src/test/java/org/infinispan/server/test/category/HotRodLocalDomain.java:
```diff
@@ -0,0 +1,13 @@
+ * @author mgencur
+ * 
+ */
+public class HotRodLocalDomain extends ClientLocal {
```
Could you tell me what does it mean a local domain?

mgencur replied:
Maybe HotRodSingleNodeDomain would be better name? It's a test with a single ISPN server (vs. Clustered where there are two servers)

slaskawi replied:
+100. HotRodSingleNodeDomain is much more descriptive.

slaskawi commented on server/integration/testsuite/src/test/java/org/infinispan/server/test/client/hotrod/AbstractRemoteCacheIT.java:
```diff
@@ -78,17 +86,54 @@ public static void release() {
+                        CustomPojoEventFilterFactory.class);
+    }
+
+    protected static Archive<?> createConverterArchive() {
```
How about extracting this to some class with deployments?

slaskawi commented on server/integration/testsuite/src/test/java/org/infinispan/server/test/client/hotrod/AbstractRemoteCacheIT.java:
```diff
@@ -78,17 +86,54 @@ public static void release() {
+    private Configuration createRemoteCacheManagerConfiguration(int... hotrodPortOverrides) {
         ConfigurationBuilder config = new ConfigurationBuilder();
+        if (hotrodPortOverrides.length != 0 && getServers().size() != hotrodPortOverrides.length) {
+            throw new IllegalArgumentException("The number of defined ports is different from server count");
```
How about using assert here? Assertions fits better into tests and this part sounds like a preconditions check.

slaskawi commented on server/integration/testsuite/src/test/java/org/infinispan/server/test/client/rest/RESTAsyncDomainIT.java:
```diff
@@ -0,0 +1,86 @@
+        try {
+            //do nothing in dist mode, the cache is already there
+            if (isReplicatedMode()) {
+                client.addSocketBinding("rest-repl", "clustered-sockets", REST_PORT1);
```
How about creating a generic initialize and cleanup methods in the abstract class (`initDistributedMode`, `initReplicatedMode` etc). This way you would only need to override some of them in your test. I think it would look a bit nicer.

mgencur replied:
Can you please explain what you exactly mean? What would it look like in this REST AsyncDomainIT class then? Thanks

slaskawi replied:
I thought about something like this:
```
class AbstractRESTAsyncIT {
...
   @BeforeClass
   public void init() {
      if(isReplicatedMode())
         initReplicatedMode();
      if(isDistributedMode())
         initDistributedMode();
   }
..
}

class RESTAsyncDomainIT {
   //No 
...
   @Override
   public void initReplicatedMode() {
      // do init here
   }
}
```
But I realized that `@BeforeClass` methods need to be static, so this won't work.

slaskawi commented on server/integration/testsuite/src/test/java/org/infinispan/server/test/client/rest/RESTClientDomainIT.java:
```diff
@@ -0,0 +1,52 @@
+import org.junit.runner.RunWith;
+
+/**
+ *
```
It would be good to put some description here.

slaskawi commented on server/integration/testsuite/src/test/java/org/infinispan/server/test/client/rest/RESTClientLocalIT.java:
```diff
@@ -8,14 +8,11 @@
- * @author <a href="mailto:jvilkola@redhat.com">Jozef Vilkolak</a>
- * @author <a href="mailto:mlinhard@redhat.com">Michal Linhard</a>
- * @version November 2013
+ * @author mgencur
```
The same here

slaskawi commented on server/integration/testsuite/src/test/java/org/infinispan/server/test/util/ManagementClient.java:
```diff
@@ -0,0 +1,221 @@
+
+import java.io.IOException;
+
+/**
```
and here

slaskawi commented on server/integration/testsuite/src/test/java/org/infinispan/server/test/util/ManagementClient.java:
```diff
@@ -0,0 +1,221 @@
+    }
+
+    public static class Constants {
+        public static final String NODE0_ADDRESS = System.getProperty("node0.ip", "127.0.0.1");
```
What is the advantage of keeping those constants in separate class? Do we want to delay their initialization? If not - I would suggest moving them to `ManagementClient`.

slaskawi commented:
Some small issues found but that's really a great piece of work!

mgencur commented:
Thanks, @slaskawi :) I'll address the issues later today.

mgencur commented:
I forgot that it is required to follow these steps to enable JMX at the beginning: http://infinispan.org/docs/8.1.x/infinispan_server_guide/infinispan_server_guide.html#_connecting_remotely_via_jmx_to_server_in_domain_mode I'll have to fix that somehow too.

mgencur commented:
Hey @slaskawi , I addressed your suggestions except moving Arq stuff to a separate class. I also added enabling/disabling jmx from the test. Feel free to squash the two commits when integrating.

slaskawi commented:
Retriggering CI

mgencur commented:
Rebased.

mgencur commented:
Rebased correctly:)

mgencur commented:
Hey. While working on adding Arquillian-Chameleon to the project, I found a little problem with two tests. This is now fixed in the latest commit of this PR. Feel free to squash all the commits when integrating.

mgencur commented:
Rebased.

danberindei commented on server/integration/testsuite/pom.xml:
```diff
@@ -269,6 +307,9 @@
         <node3.ip>127.0.0.1</node3.ip>
+        <node0.mgmt.port>9990</node0.mgmt.port>
+        <node1.mgmt.port>10090</node1.mgmt.port>
+        <node2.mgmt.port>10190</node2.mgmt.port>
```
missing `node3.mgmt.port`?

mgencur replied:
node3 is only used in tests for rolling upgrades. These are not run as part of domain mode tests. We can add it later when we modify rolling ugprade tests for domain mode.

danberindei replied:
Ok, it just seemed a bit odd to have `node3.ip` without a port.

mgencur commented:
Can someone please merge this PR?

tristantarrant commented:
Pulling...

tristantarrant commented:
[ERROR] Failed to execute goal on project infinispan-server-testsuite: Could not resolve dependencies for project org.infinispan.server:infinispan-server-testsuite:jar:9.0.0-SNAPSHOT: The following artifacts could not be resolved: org.wildfly.extras.creaper:creaper-core:jar:0.9.5, org.wildfly.extras.creaper:creaper-commands:jar:0.9.5: Failure to find org.wildfly.extras.creaper:creaper-core:jar:0.9.5 in https://repository.jboss.org/nexus/content/groups/public-jboss/ was cached in the local repository, resolution will not be reattempted until the update interval of jboss-public-repository-group has elapsed or updates are forced -> [Help 1]

mgencur commented:
The artifacts can be found at http://repository.jboss.org/nexus/content/groups/public/org/wildfly/extras/creaper/creaper-core/0.9.5/  Not sure what is the purpose of "public-jboss" repository which doesn't have it.

mgencur commented:
Rebased.

tristantarrant commented:
Can you please also amend maven-settings.xml pointing to the correct repo ?

mgencur commented:
Done.

tristantarrant commented:
Pushed, thanks @mgencur
